src/traccuracy/matchers/_base.py

```diff
@@ -111,6 +112,36 @@ def __init__(
         self.gt_pred_map = dict(gt_pred_map)
         self.pred_gt_map = dict(pred_gt_map)
 
+        self._matching_type = None
+
+    @property
+    def matching_type(self):
+        """Determines the matching type: one-to-one, one-to-many, many-to-one, many-to-many"""
+        if len(self.mapping) == 0:
+            warnings.warn(
+                "Mapping is empty. Defaulting to type of one-to-one", stacklevel=2
+            )
+
+        if self._matching_type is not None:
+            return self._matching_type
+
+        gt_count = Counter(m[0] for m in self.mapping)
```

Comment:
I think we could get away with a slightly cheaper computation here by iterating through the `gt_pred_map` and `pred_gt_map` directly, and breaking when we hit a duplicate. Something like

```python
gt_type = "one"
for label_id, matches in self.gt_pred_map.items():
   if len(matches) > 1:
       gt_type = "many"
       break
```

And then the same for the `pred_type`. In the worst case we still make one full iteration of each, but we save on creating the `Counter` object and then casting to `set`

Reply:
That's much more elegant!